runtime/lua/vim/iter.lua

```diff
@@ -18,10 +18,13 @@ ListIter.__call = function(self)
   return self:next()
 end
 
+--- Packed tables use this as their metatable
+local packedmt = {}
+
 ---@private
 local function unpack(t)
-  if type(t) == 'table' and t.__n ~= nil then
-    return _G.unpack(t, 1, t.__n)
+  if getmetatable(t) == packedmt then
+    return _G.unpack(t)
```

Comment:
Is this support to work on arrays with holes?

Reply:
Yes, we do need to support holes because a function in the iterator pipeline could return multiple values, one of which is `nil`. I'm thinking about the best way to handle this though.

The current implementation (without this PR) already handles this because the `__n` tag doubles as the length indicator. However, when the iterator is serialized into a table we strip the `__n` tags. I'm wondering if we should leave those in.

Basically, consider the following scenario:

```lua
vim.iter({1, 2, 3}):map(function(v)
  return v, nil, v*v
end):totable()
```

What should this return?

Status quo has the whole table, but the `__n` tag is removed:

```lua
{ { 1, [3] = 1 }, { 2, [3] = 4 }, { 3, [3] = 9 } }
```

I'm wondering if we should rename `__n` to just `n` and leave it in the resulting table, which preserves full information on the table data, but then adds an extra field that may surprise users:

```lua
{ { n = 3, 1, [3] = 1 }, { n = 3, 2, [3] = 4 }, { n = 3, 3, [3] = 9 } }
```

Reply:
I'd keep `__n` omitted. 

> Or we just leave holes in the returned table 

This seems like the obvious choice to me.